File: eurorack_case/parts.py
"""Parts of the asymmetric Eurorack case.

Frame: X runs along the rails (0 .. width), Y runs front (-) to rear (+) and
is centred on the panel, Z is up with the top of the walls at Z = 0. A module
panel sits in the 2 mm recess between Z = -2 and Z = 0.

Parts:
  case()        the shell -- sloped floor, front/rear walls, T-slot rails,
                open at both ends, screw holes for the end caps
  end_cap()     4 mm plate closing one end, with a 5 mm tab that plugs into
                the shell and carries M3 nut traps
  blank_panel() a blank module panel, to check the rail spacing
"""
from math import hypot

import cadquery as cq

from .params import HP, CaseParams

EPS = 1e-3


class Floor:
    """The sloped bottom plate, as the line joining the two wall feet."""

    def __init__(self, p: CaseParams):
        self.y0, self.z0 = -p.outer_half, -p.front_height
        self.y1, self.z1 = p.outer_half, -p.rear_height
        dy, dz = self.y1 - self.y0, self.z1 - self.z0
        length = hypot(dy, dz)
        self.slope = dz / dy
        self.sec = length / dy                 # 1 / cos(tilt)
        self.inward = cq.Vector(0, -dz / length, dy / length)

    def z_outer(self, y: float) -> float:
        return self.z0 + (y - self.y0) * self.slope

    def z_at(self, y: float, offset: float) -> float:
        """Z of the line `offset` mm inside the outer floor surface, at Y = y."""
        return self.z_outer(y) + offset * self.sec

    def outer_plane(self) -> cq.Plane:
        """Workplane on the outside of the floor. Local x = X, normal points out."""
        return cq.Plane(
            origin=(0, 0, self.z_outer(0)),
            xDir=(1, 0, 0),
            normal=-self.inward,
        )


def front_plane(p: CaseParams) -> cq.Plane:
    return cq.Plane(origin=(0, -p.outer_half, 0), xDir=(1, 0, 0), normal=(0, -1, 0))


def rear_plane(p: CaseParams) -> cq.Plane:
    return cq.Plane(origin=(0, p.outer_half, 0), xDir=(1, 0, 0), normal=(0, 1, 0))


def _profile(points, length: float, x0: float = 0.0) -> cq.Workplane:
    """Extrude a closed YZ polygon from X = x0 by `length` along +X."""
    wp = cq.Workplane("YZ", origin=(x0, 0, 0))
    return wp.polyline(points).close().extrude(length)


def long_edges(body: cq.Workplane, yz_points, tol: float = 0.05) -> cq.Workplane:
    """Edges parallel to X whose centre lies at one of the given (Y, Z) points.

    This replaces Onshape's history-based edge references: the lip edges are
    known by where they are, so select them by where they are.
    """
    wanted, hit = [], set()
    for e in body.edges("|X").vals():
        c = e.Center()
        for i, (y, z) in enumerate(yz_points):
            if abs(c.y - y) < tol and abs(c.z - z) < tol:
                wanted.append(e)
                hit.add(i)
    if len(hit) != len(yz_points):
        raise ValueError(f"expected lip edges at {len(yz_points)} positions, found {len(hit)}")
    return body.newObject(wanted)


def apply_lips(body: cq.Workplane, p: CaseParams, fl: Floor) -> cq.Workplane:
    """Round / chamfer / leave the four long outer edges, per the configuration."""
    yo = p.outer_half
    top = [(-yo, 0.0), (yo, 0.0)]
    bottom = [(-yo, -p.front_height), (yo, -p.rear_height)]
    for style, size_round, size_chamfer, edges in (
        (p.top_lips, p.upper_round, p.upper_chamfer, top),
        (p.bottom_lips, p.lower_round, p.lower_chamfer, bottom),
    ):
        if style == "round":
            body = long_edges(body, edges).fillet(size_round)
        elif style == "chamfer":
            body = long_edges(body, edges).chamfer(size_chamfer)
    return body


# --------------------------------------------------------------------------
# screws that hold the end caps on
# --------------------------------------------------------------------------

def wall_hole_zs(p: CaseParams, fl: Floor, height: float, y_leg_inner: float):
    """Z of the screws through a wall of the given outer height.

    The source puts the first screw at mid-height and patterns the rest at
    `wall_hole_spacing`. Keep every position whose nut trap fits in the tab's
    leg -- between the underside of the rail and the tab's floor -- and whose
    counterbore stays on the wall.
    """
    half_pocket = p.nut_pocket_width / 2
    half_cb = p.bolt_counterbore_diam / 2
    leg_top = p.rail_bottom
    leg_floor = fl.z_at(y_leg_inner, p.bottom_thickness + p.tab_clearance + p.end_cap_tab_thickness)
    zs = []
    for k in range(-8, 9):
        z = -height / 2 + k * p.wall_hole_spacing
        if z + half_pocket > leg_top + EPS or z - half_pocket < leg_floor - EPS:
            continue
        if z + half_cb > 0 or z - half_cb < -height:
            continue
        zs.append(z)
    return sorted(zs)


def _screw_set(plane: cq.Plane, world_points, p: CaseParams, into: float) -> cq.Workplane:
    """Counterbore + through hole + nut trap for screws entering at `world_points`.

    `into` is the distance from the outer surface to the tab's inner face,
    i.e. where the nut trap opens.
    """
    pts = [plane.toLocalCoords(cq.Vector(*w)) for w in world_points]
    pts = [(v.x, v.y) for v in pts]
    wp = cq.Workplane(plane)
    counterbore = wp.pushPoints(pts).circle(p.bolt_counterbore_diam / 2).extrude(-p.bolt_counterbore_depth)
    through = wp.pushPoints(pts).circle(p.bolt_through_diam / 2).extrude(-(into + 1.0))
    pocket = (
        wp.workplane(offset=-into)
        .pushPoints(pts)
        .rect(p.nut_pocket_width, p.nut_pocket_width)
        .extrude(p.rail_nut_depth)
    )
    return counterbore.union(through).union(pocket)


def screw_cutters(p: CaseParams, fl: Floor, x: float) -> cq.Workplane:
    """Everything to subtract for the screws at one end of the case (axis at X = x)."""
    yi, yo = p.inner_half, p.outer_half
    tt, c = p.end_cap_tab_thickness, p.tab_clearance
    into = p.wall_thickness + c + tt

    # floor: three screws spaced along the tilted surface, centred on Y = 0
    fplane = fl.outer_plane()
    floor_pts = [(x, s) for s in (-p.bottom_hole_spacing, 0.0, p.bottom_hole_spacing)]
    floor_pts = [fplane.toWorldCoords((lx, ly)).toTuple() for lx, ly in floor_pts]
    cutters = _screw_set(fplane, floor_pts, p, into)

    # walls
    for plane, height, y_out in (
        (front_plane(p), p.front_height, -yo),
        (rear_plane(p), p.rear_height, yo),
    ):
        y_leg_inner = (yi - c - tt) * (1 if y_out > 0 else -1)
        zs = wall_hole_zs(p, fl, height, y_leg_inner)
        if zs:
            pts = [(x, y_out, z) for z in zs]
            cutters = cutters.union(_screw_set(plane, pts, p, into))
    return cutters


def all_screw_cutters(p: CaseParams, fl: Floor) -> cq.Workplane:
    """Screw sets for every end that takes a removable cap."""
    right = screw_cutters(p, fl, p.width - p.bottom_hole_inset)
    if "left" in p.closed_ends:
        return right
    return screw_cutters(p, fl, p.bottom_hole_inset).union(right)


# --------------------------------------------------------------------------
# parts
# --------------------------------------------------------------------------

def shell_profile(p: CaseParams, fl: Floor):
    yi, yo = p.inner_half, p.outer_half
    t = p.bottom_thickness
    return [
        (-yo, 0.0), (-yo, -p.front_height), (yo, -p.rear_height), (yo, 0.0),
        (yi, 0.0), (yi, fl.z_at(yi, t)), (-yi, fl.z_at(-yi, t)), (-yi, 0.0),
    ]


def case(p: CaseParams = CaseParams()) -> cq.Workplane:
    fl = Floor(p)
    w = p.width
    body = _profile(shell_profile(p, fl), w)

    # T-slot rails along both inner walls, under the panel recess
    yi = p.inner_half
    rail_cz = (p.rail_top + p.rail_bottom) / 2
    slot_top = p.rail_top + 1.0                      # break cleanly into the recess
    slot_bottom = p.rail_top - p.rail_bolt_depth
    nut_top = p.rail_top - p.rail_nut_depth
    nut_bottom = nut_top - p.rail_nut_thickness
    for s in (1, -1):
        rail = (
            cq.Workplane("YZ")
            .center(s * (yi - p.rail_depth / 2), rail_cz)
            .rect(p.rail_depth, p.rail_height)
            .extrude(w)
        )
        body = body.union(rail)
        axis = s * (yi - p.rail_inset)
        bolt_slot = (
            cq.Workplane("YZ")
            .center(axis, (slot_top + slot_bottom) / 2)
            .rect(p.rail_bolt_width, slot_top - slot_bottom)
            .extrude(w)
        )
        nut_slot = (
            cq.Workplane("YZ")
            .center(axis, (nut_top + nut_bottom) / 2)
            .rect(p.rail_nut_slot_width, nut_top - nut_bottom)
            .extrude(w)
        )
        body = body.cut(bolt_slot).cut(nut_slot)

    if p.left_wall:
        # the "asym" variant: a plate with the outer profile closes X = 0 and the lips
        # run along it uninterrupted. Unlike a cap, its outside face is not edge-broken.
        wall = _profile(end_plate_profile(p), p.end_cap_thickness, x0=-p.end_cap_thickness)
        body = body.union(wall)

    body = apply_lips(body, p, fl)
    return body.cut(all_screw_cutters(p, fl))


def end_plate_profile(p: CaseParams):
    yo = p.outer_half
    return [(-yo, 0.0), (-yo, -p.front_height), (yo, -p.rear_height), (yo, 0.0)]


def tab_profile(p: CaseParams, fl: Floor):
    """U-shaped band lining the cavity below the rails."""
    c, tt, bt = p.tab_clearance, p.end_cap_tab_thickness, p.bottom_thickness
    yo_ = p.inner_half - c            # outside of the tab
    yi_ = yo_ - tt                    # inside of the tab
    top = p.rail_bottom - c
    return [
        (-yo_, top), (-yo_, fl.z_at(-yo_, bt + c)), (yo_, fl.z_at(yo_, bt + c)), (yo_, top),
        (yi_, top), (yi_, fl.z_at(yi_, bt + c + tt)), (-yi_, fl.z_at(-yi_, bt + c + tt)), (-yi_, top),
    ]


def end_cap(p: CaseParams = CaseParams(), end: str = "left") -> cq.Workplane:
    """End cap in its assembled position. `left` closes X = 0, `right` closes X = width."""
    if end in p.closed_ends:
        raise ValueError(f"the {end} end is an integral wall in this configuration")
    fl = Floor(p)
    plate = _profile(end_plate_profile(p), p.end_cap_thickness, x0=-p.end_cap_thickness)
    plate = plate.faces("<X").chamfer(p.end_cap_face_chamfer)
    plate = apply_lips(plate, p, fl)
    tab = _profile(tab_profile(p, fl), p.end_cap_tab_depth)
    cap = plate.union(tab).cut(screw_cutters(p, fl, p.bottom_hole_inset))
    if end == "right":
        cap = cap.mirror("YZ", basePointVector=(p.width / 2, 0, 0))
    elif end != "left":
        raise ValueError(end)
    return cap


def blank_panel(p: CaseParams = CaseParams(), hp: int = 6, x0: float = 0.0) -> cq.Workplane:
    """A blank module panel `hp` wide, sitting in the recess from X = x0."""
    w = hp * HP
    yi = p.inner_half
    panel = (
        cq.Workplane("XY", origin=(x0, 0, 0))
        .center(w / 2, 0)
        .rect(w, p.panel_height)
        .extrude(-p.panel_thickness)
    )
    xs = (x0 + p.panel_hole_side_inset, x0 + w - p.panel_hole_side_inset)
    ys = (-(yi - p.panel_hole_edge_inset), yi - p.panel_hole_edge_inset)
    holes = (
        cq.Workplane("XY")
        .pushPoints([(x, y) for x in xs for y in ys])
        .circle(p.panel_hole_diam / 2)
        .extrude(-p.panel_thickness)
    )
    return panel.cut(holes)


def assembly(p: CaseParams = CaseParams(), panel_hp: int = 6) -> cq.Assembly:
    a = cq.Assembly(name="eurorack_case")
    a.add(case(p), name="case", color=cq.Color(0.55, 0.55, 0.6))
    for end in ("left", "right"):
        if end not in p.closed_ends:
            a.add(end_cap(p, end), name=f"end_cap_{end}", color=cq.Color(0.8, 0.5, 0.2))
    a.add(blank_panel(p, panel_hp), name="blank_panel", color=cq.Color(0.85, 0.85, 0.85))
    return a

File: web/src/model/parts.ts
/**
 * Parts of the Eurorack case in replicad — a port of eurorack_case/parts.py.
 *
 * Frame: X runs along the rails (0 .. width), Y runs front (-) to rear (+) and
 * is centred on the panel, Z is up with the top of the walls at Z = 0.
 *
 *   caseShell()   the shell: sloped floor, front/rear walls, T-slot rails,
 *                 screw holes for the end caps; open at both ends, or closed
 *                 at X = 0 by an integral wall (leftWall)
 *   endCap()      4 mm plate closing one end, with a 5 mm tab that plugs into
 *                 the shell and carries M3 nut traps
 *   blankPanel()  a blank module panel, to check the rail spacing
 */
import {
  draw,
  drawCircle,
  drawRectangle,
  makeCylinder,
  Plane,
  type AnyShape,
  type Drawing,
  type EdgeFinder,
  type Shape3D,
} from "replicad";

import { HP, caseParams, type CaseParams, type CaseParamsInput } from "./params";

type Pt = [number, number];
type Vec3 = [number, number, number];

const EPS = 1e-3;

/** replicad types a sketch extrusion as any shape; ours are always solids. */
const solid = (s: AnyShape): Shape3D => s as Shape3D;

const add = (a: Vec3, b: Vec3, k = 1): Vec3 => [a[0] + b[0] * k, a[1] + b[1] * k, a[2] + b[2] * k];
const neg = (a: Vec3): Vec3 => [-a[0], -a[1], -a[2]];

/** The sloped bottom plate, as the line joining the two wall feet. */
export class Floor {
  readonly y0: number; readonly z0: number; readonly y1: number; readonly z1: number;
  readonly slope: number;
  readonly sec: number;           // 1 / cos(tilt)
  readonly inward: Vec3;          // unit normal pointing into the case

  constructor(p: CaseParams) {
    this.y0 = -p.outerHalf; this.z0 = -p.frontHeight;
    this.y1 = p.outerHalf; this.z1 = -p.rearHeight;
    const dy = this.y1 - this.y0, dz = this.z1 - this.z0;
    const length = Math.hypot(dy, dz);
    this.slope = dz / dy;
    this.sec = length / dy;
    this.inward = [0, -dz / length, dy / length];
  }

  zOuter(y: number): number { return this.z0 + (y - this.y0) * this.slope; }

  /** Z of the line `offset` mm inside the outer floor surface, at Y = y. */
  zAt(y: number, offset: number): number { return this.zOuter(y) + offset * this.sec; }

  /** Workplane on the outside of the floor. Local x = X, normal points out. */
  outerPlane(): Plane { return new Plane([0, 0, this.zOuter(0)], [1, 0, 0], neg(this.inward)); }
}

export const frontPlane = (p: CaseParams): Plane => new Plane([0, -p.outerHalf, 0], [1, 0, 0], [0, -1, 0]);
export const rearPlane = (p: CaseParams): Plane => new Plane([0, p.outerHalf, 0], [1, 0, 0], [0, 1, 0]);

/** A closed (Y, Z) polygon, extruded from X = x0 by `length` along +X. */
function profile(points: Pt[], length: number, x0 = 0): Shape3D {
  const pen = draw(points[0]);
  for (const pt of points.slice(1)) pen.lineTo(pt);
  return solid(pen.close().sketchOnPlane("YZ", x0).extrude(length));
}

/** A (Y, Z) rectangle centred at (cy, cz), extruded along X. */
function bar(cy: number, cz: number, w: number, h: number, length: number): Shape3D {
  return solid(drawRectangle(w, h).translate(cy, cz).sketchOnPlane("YZ").extrude(length));
}

/**
 * Edges parallel to X lying on the line through (Y, Z) = (y, z).
 *
 * Selecting by position replaces Onshape's history-based references. A thin
 * box along the whole line also catches collinear pieces, which is what the
 * integral wall's join leaves behind.
 */
function longEdges(f: EdgeFinder, yz: Pt[], xMin: number, xMax: number): EdgeFinder {
  return f.inDirection("X").either(
    yz.map(([y, z]) => (g: EdgeFinder) => g.inBox([xMin - 1, y - 0.05, z - 0.05], [xMax + 1, y + 0.05, z + 0.05])),
  );
}

/** Round / chamfer / leave the four long outer edges, per the configuration. */
function applyLips(body: Shape3D, p: CaseParams, xMin: number, xMax: number): Shape3D {
  const yo = p.outerHalf;
  const top: Pt[] = [[-yo, 0], [yo, 0]];
  const bottom: Pt[] = [[-yo, -p.frontHeight], [yo, -p.rearHeight]];
  const passes: [string, number, number, Pt[]][] = [
    [p.topLips, p.upperRound, p.upperChamfer, top],
    [p.bottomLips, p.lowerRound, p.lowerChamfer, bottom],
  ];
  for (const [style, round, chamfer, edges] of passes) {
    if (style === "round") body = body.fillet(round, (e) => longEdges(e, edges, xMin, xMax));
    else if (style === "chamfer") body = body.chamfer(chamfer, (e) => longEdges(e, edges, xMin, xMax));
  }
  return body;
}

// --------------------------------------------------------------------------
// screws that hold the end caps on
// --------------------------------------------------------------------------

/**
 * Z of the screws through a wall of the given outer height.
 *
 * The source puts the first screw at mid-height and patterns the rest at
 * `wallHoleSpacing`. Keep every position whose nut trap fits in the tab's leg
 * — between the underside of the rail and the tab's floor — and whose
 * counterbore stays on the wall.
 */
export function wallHoleZs(p: CaseParams, fl: Floor, height: number, yLegInner: number): number[] {
  const halfPocket = p.nutPocketWidth / 2, halfCb = p.boltCounterboreDiam / 2;
  const legTop = p.railBottom;
  const legFloor = fl.zAt(yLegInner, p.bottomThickness + p.tabClearance + p.endCapTabThickness);
  const zs: number[] = [];
  for (let k = -8; k <= 8; k++) {
    const z = -height / 2 + k * p.wallHoleSpacing;
    if (z + halfPocket > legTop + EPS || z - halfPocket < legFloor - EPS) continue;
    if (z + halfCb > 0 || z - halfCb < -height) continue;
    zs.push(z);
  }
  return zs.sort((a, b) => a - b);
}

interface ScrewSite { point: Vec3; normal: Vec3 }

/** Where each screw enters the shell at one end (axis at X = x), with the outward normal. */
function screwSites(p: CaseParams, fl: Floor, x: number): ScrewSite[] {
  const yi = p.innerHalf, yo = p.outerHalf, tt = p.endCapTabThickness, c = p.tabClearance;
  const sites: ScrewSite[] = [];
  const floor = fl.outerPlane();
  for (const s of [-p.bottomHoleSpacing, 0, p.bottomHoleSpacing]) {
    const w = floor.toWorldCoords([x, s, 0]);
    sites.push({ point: [w.x, w.y, w.z], normal: neg(fl.inward) });
  }
  for (const z of wallHoleZs(p, fl, p.frontHeight, -(yi - c - tt))) sites.push({ point: [x, -yo, z], normal: [0, -1, 0] });
  for (const z of wallHoleZs(p, fl, p.rearHeight, yi - c - tt)) sites.push({ point: [x, yo, z], normal: [0, 1, 0] });
  return sites;
}

/**
 * Counterbore + through hole + nut trap for one screw. `into` is the distance
 * from the outer surface to the tab's inner face, where the trap opens.
 */
function screwCutters(site: ScrewSite, p: CaseParams, into: number): Shape3D[] {
  const { point, normal } = site;
  const counterbore = makeCylinder(p.boltCounterboreDiam / 2, p.boltCounterboreDepth, point, neg(normal));
  const through = makeCylinder(p.boltThroughDiam / 2, into + 1, point, neg(normal));
  const pocketPlane = new Plane(add(point, normal, -into), [1, 0, 0], normal);
  const pocket = solid(drawRectangle(p.nutPocketWidth, p.nutPocketWidth).sketchOnPlane(pocketPlane).extrude(p.railNutDepth));
  return [counterbore, through, pocket];
}

function cutAll(body: Shape3D, tools: Shape3D[]): Shape3D {
  for (const t of tools) body = body.cut(t);
  return body;
}

function endScrewCutters(p: CaseParams, fl: Floor, x: number): Shape3D[] {
  const into = p.wallThickness + p.tabClearance + p.endCapTabThickness;
  return screwSites(p, fl, x).flatMap((s) => screwCutters(s, p, into));
}

/** Screw sets for every end that takes a removable cap. */
function allScrewCutters(p: CaseParams, fl: Floor): Shape3D[] {
  const right = endScrewCutters(p, fl, p.width - p.bottomHoleInset);
  if (p.closedEnds.includes("left")) return right;
  return [...endScrewCutters(p, fl, p.bottomHoleInset), ...right];
}

// --------------------------------------------------------------------------
// parts
// --------------------------------------------------------------------------

function shellProfile(p: CaseParams, fl: Floor): Pt[] {
  const yi = p.innerHalf, yo = p.outerHalf, t = p.bottomThickness;
  return [
    [-yo, 0], [-yo, -p.frontHeight], [yo, -p.rearHeight], [yo, 0],
    [yi, 0], [yi, fl.zAt(yi, t)], [-yi, fl.zAt(-yi, t)], [-yi, 0],
  ];
}

function endPlateProfile(p: CaseParams): Pt[] {
  const yo = p.outerHalf;
  return [[-yo, 0], [-yo, -p.frontHeight], [yo, -p.rearHeight], [yo, 0]];
}

/** U-shaped band lining the cavity below the rails. */
function tabProfile(p: CaseParams, fl: Floor): Pt[] {
  const c = p.tabClearance, tt = p.endCapTabThickness, bt = p.bottomThickness;
  const yo = p.innerHalf - c;      // outside of the tab
  const yi = yo - tt;              // inside of the tab
  const top = p.railBottom - c;
  return [
    [-yo, top], [-yo, fl.zAt(-yo, bt + c)], [yo, fl.zAt(yo, bt + c)], [yo, top],
    [yi, top], [yi, fl.zAt(yi, bt + c + tt)], [-yi, fl.zAt(-yi, bt + c + tt)], [-yi, top],
  ];
}

export function caseShell(input: CaseParamsInput | CaseParams = {}): Shape3D {
  const p = "width" in input ? input : caseParams(input);
  const fl = new Floor(p);
  const w = p.width;
  let body = profile(shellProfile(p, fl), w);

  // T-slot rails along both inner walls, under the panel recess
  const yi = p.innerHalf;
  const railCz = (p.railTop + p.railBottom) / 2;
  const slotTop = p.railTop + 1;                 // break cleanly into the recess
  const slotBottom = p.railTop - p.railBoltDepth;
  const nutTop = p.railTop - p.railNutDepth;
  const nutBottom = nutTop - p.railNutThickness;
  for (const s of [1, -1]) {
    body = body.fuse(bar(s * (yi - p.railDepth / 2), railCz, p.railDepth, p.railHeight, w));
    const axis = s * (yi - p.railInset);
    body = body.cut(bar(axis, (slotTop + slotBottom) / 2, p.railBoltWidth, slotTop - slotBottom, w));
    body = body.cut(bar(axis, (nutTop + nutBottom) / 2, p.railNutSlotWidth, nutTop - nutBottom, w));
  }

  let xMin = 0;
  if (p.leftWall) {
    // the "asym" variant: a plate with the outer profile closes X = 0 and the lips
    // run along it uninterrupted. Unlike a cap, its outside face is not edge-broken.
    body = body.fuse(profile(endPlateProfile(p), p.endCapThickness, -p.endCapThickness));
    xMin = -p.endCapThickness;
  }

  body = applyLips(body, p, xMin, w);
  return cutAll(body, allScrewCutters(p, fl));
}

/** End cap in its assembled position. `left` closes X = 0, `right` closes X = width. */
export function endCap(input: CaseParamsInput | CaseParams = {}, end: "left" | "right" = "left"): Shape3D {
  const p = "width" in input ? input : caseParams(input);
  if (p.closedEnds.includes(end)) throw new Error(`the ${end} end is an integral wall in this configuration`);
  const fl = new Floor(p);
  const x0 = -p.endCapThickness;
  let plate = profile(endPlateProfile(p), p.endCapThickness, x0);
  plate = plate.chamfer(p.endCapFaceChamfer, (e) => e.inPlane("YZ", x0));
  plate = applyLips(plate, p, x0, 0);
  const tab = profile(tabProfile(p, fl), p.endCapTabDepth);
  let cap = cutAll(plate.fuse(tab), endScrewCutters(p, fl, p.bottomHoleInset));
  if (end === "right") cap = cap.mirror("YZ", [p.width / 2, 0, 0]);
  return cap;
}

/** A blank module panel `hp` wide, sitting in the recess from X = x0. */
export function blankPanel(input: CaseParamsInput | CaseParams = {}, hp = 6, x0 = 0): Shape3D {
  const p = "width" in input ? input : caseParams(input);
  const w = hp * HP, yi = p.innerHalf;
  let panel: Drawing = drawRectangle(w, p.panelHeight).translate(x0 + w / 2, 0);
  for (const x of [x0 + p.panelHoleSideInset, x0 + w - p.panelHoleSideInset]) {
    for (const y of [-(yi - p.panelHoleEdgeInset), yi - p.panelHoleEdgeInset]) {
      panel = panel.cut(drawCircle(p.panelHoleDiam / 2).translate(x, y));
    }
  }
  return solid(panel.sketchOnPlane("XY", -p.panelThickness).extrude(p.panelThickness));
}
